File: android/app/src/main/java/com/openburnbar/data/policy/MobileMercuryMediaPolicy.kt
package com.openburnbar.data.policy

enum class MobileMercuryCapability(val raw: String) {
    MIRROR_VIEWER("mirror.viewer"),
    MIRROR_HOST("mirror.host"),
    MIRROR_AUTO_ACCEPT("mirror.auto_accept"),
    REMOTE_UNLOCK_HOST("remote_unlock.host"),
    FILE_SEND("file.send"),
    FILE_RECEIVE("file.receive"),
    CALL_RECEIVE("call.receive"),
    CALL_ORIGINATE("call.originate"),
    ;

    companion object {
        fun fromRaw(raw: String): MobileMercuryCapability? = entries.firstOrNull { it.raw == raw }
    }
}

enum class MobileMercuryInviteAck(val wire: String) {
    PAIRED("paired"),
    MISMATCH("mismatch"),
    DENIED("denied"),
}

enum class MobileMercurySessionPresentation(val wire: String) {
    IDLE("idle"),
    CONNECTED("connected"),
    RECONNECTING("reconnecting"),
    DENIED("denied"),
    FAILED("failed"),
}

/** Mercury/media.control decisions. Source: iOS MercuryPeer + 60s heartbeat. */
object MobileMercuryMediaPolicy {
    const val HEARTBEAT_INTERVAL_MS: Long = 60_000L

    val phoneHeartbeatCapabilities: List<String> = listOf(
        MobileMercuryCapability.MIRROR_VIEWER.raw,
        MobileMercuryCapability.FILE_SEND.raw,
        MobileMercuryCapability.FILE_RECEIVE.raw,
        MobileMercuryCapability.CALL_RECEIVE.raw,
    )

    fun filterCapabilities(raw: Collection<String>): List<String> {
        val known = MobileMercuryCapability.entries.map { it.raw }.toSet()
        return raw.filter { known.contains(it) }.toSet().sorted()
    }

    fun inviteAckPair(inviteId: String, ackId: String, accepted: Boolean): MobileMercuryInviteAck {
        if (inviteId.trim().isEmpty() || ackId.trim().isEmpty() || inviteId != ackId) {
            return MobileMercuryInviteAck.MISMATCH
        }
        return if (accepted) MobileMercuryInviteAck.PAIRED else MobileMercuryInviteAck.DENIED
    }

    fun sessionPresentation(phase: String, denied: Boolean): MobileMercurySessionPresentation {
        if (denied) return MobileMercurySessionPresentation.DENIED
        return when (phase) {
            "live" -> MobileMercurySessionPresentation.CONNECTED
            "reconnecting" -> MobileMercurySessionPresentation.RECONNECTING
            "failed" -> MobileMercurySessionPresentation.FAILED
            else -> MobileMercurySessionPresentation.IDLE
        }
    }

    fun canRequestMirror(isOnline: Boolean, capabilities: Collection<String>): Boolean =
        isOnline && capabilities.contains(MobileMercuryCapability.MIRROR_HOST.raw)

    fun canPlaceCall(isOnline: Boolean, capabilities: Collection<String>): Boolean = isOnline && capabilities.contains(MobileMercuryCapability.CALL_RECEIVE.raw)

    fun canSendFile(isOnline: Boolean, capabilities: Collection<String>): Boolean = isOnline && capabilities.contains(MobileMercuryCapability.FILE_RECEIVE.raw)
}

File: OpenBurnBarCore/Sources/OpenBurnBarKernel/SharedModels/MobileMercuryMediaPolicy.swift
import Foundation

/// Shared Mercury capability tokens. Must match `MercuryPeer.Feature`.
public enum MobileMercuryCapability: String, Sendable, Equatable, CaseIterable {
    case mirrorViewer = "mirror.viewer"
    case mirrorHost = "mirror.host"
    case mirrorAutoAccept = "mirror.auto_accept"
    case remoteUnlockHost = "remote_unlock.host"
    case fileSend = "file.send"
    case fileReceive = "file.receive"
    case callReceive = "call.receive"
    case callOriginate = "call.originate"
}

public enum MobileMercuryInviteAck: String, Sendable, Equatable {
    case paired
    case mismatch
    case denied
}

public enum MobileMercurySessionPresentation: String, Sendable, Equatable {
    case idle
    case connected
    case reconnecting
    case denied
    case failed
}

/// Mercury/media.control decisions. Source: iOS `MercuryPeer` + 60s heartbeat.
public enum MobileMercuryMediaPolicy {
    public static let heartbeatIntervalMs: Int64 = 60_000

    public static let phoneHeartbeatCapabilities: [String] = [
        MobileMercuryCapability.mirrorViewer.rawValue,
        MobileMercuryCapability.fileSend.rawValue,
        MobileMercuryCapability.fileReceive.rawValue,
        MobileMercuryCapability.callReceive.rawValue
    ]

    public static func filterCapabilities(_ raw: [String]) -> [String] {
        let known = Set(MobileMercuryCapability.allCases.map(\.rawValue))
        return Array(Set(raw.filter { known.contains($0) })).sorted()
    }

    public static func inviteAckPair(
        inviteId: String,
        ackId: String,
        accepted: Bool
    ) -> MobileMercuryInviteAck {
        if inviteId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            || ackId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            || inviteId != ackId {
            return .mismatch
        }
        return accepted ? .paired : .denied
    }

    /// Permission denial is never a connected session.
    public static func sessionPresentation(
        phase: String,
        denied: Bool
    ) -> MobileMercurySessionPresentation {
        if denied { return .denied }
        switch phase {
        case "live":
            return .connected
        case "reconnecting":
            return .reconnecting
        case "failed":
            return .failed
        default:
            return .idle
        }
    }

    public static func canRequestMirror(isOnline: Bool, capabilities: [String]) -> Bool {
        isOnline && capabilities.contains(MobileMercuryCapability.mirrorHost.rawValue)
    }

    public static func canPlaceCall(isOnline: Bool, capabilities: [String]) -> Bool {
        isOnline && capabilities.contains(MobileMercuryCapability.callReceive.rawValue)
    }

    public static func canSendFile(isOnline: Bool, capabilities: [String]) -> Bool {
        isOnline && capabilities.contains(MobileMercuryCapability.fileReceive.rawValue)
    }
}
